subprojects/snapshots/src/integTest/groovy/org/gradle/internal/vfs/VirtualFileSystemRetentionIntegrationTest.groovy

```diff
@@ -0,0 +1,198 @@
+
+        when:
+        taskSourceFile.text = taskWithGreeting("Hello from modified task!")
+        withRetentionRun "hello"
+        then:
+        outputContains "Hello from original task!"
+
+        when:
+        withRetentionRun "hello", "-D${VFS_CHANGES_SINCE_LAST_BUILD_PROPERTY}=${taskSourceFile.absolutePath}"
+        then:
+        outputContains "Hello from modified task!"
+    }
+
+    def "build script changes get recognized"() {
+        when:
+        buildFile.text = """
+            println "Hello from the build!"
+        """
+        withRetentionRun "help"
+        then:
+        outputContains "Hello from the build!"
+
+        when:
+        buildFile.text = """
+            println "Hello from the modified build!"
+        """
+        withRetentionRun "help"
+        then:
+        outputContains "Hello from the modified build!"
+    }
+
+    def "settings script changes get recognized"() {
+        when:
+        settingsFile.text = """
+            println "Hello from settings!"
+        """
+        withRetentionRun "help"
+        then:
+        outputContains "Hello from settings!"
+
+        when:
+        settingsFile.text = """
+            println "Hello from modified settings!"
+        """
+        withRetentionRun "help"
+        then:
+        outputContains "Hello from modified settings!"
+    }
+
+    def "source file changes are recognized when retention has just been enabled"() {
+        buildFile << """
+            apply plugin: "application"
+            
+            application.mainClassName = "Main"
+        """
+
+        def mainSourceFile = file("src/main/java/Main.java")
+        mainSourceFile.text = sourceFileWithGreeting("Hello World!")
+
+        when:
+        withoutRetentionRun "run"
+        then:
+        outputContains "Hello World!"
+        executedAndNotSkipped ":compileJava", ":classes", ":run"
+
+        when:
+        mainSourceFile.text = sourceFileWithGreeting("Hello VFS!")
+        withRetentionRun "run"
+        then:
+        outputContains "Hello VFS!"
+        executedAndNotSkipped ":compileJava", ":classes", ":run"
+    }
+
+    def "source file changes are recognized when retention has just been disabled"() {
+        buildFile << """
+            apply plugin: "application"
+            
+            application.mainClassName = "Main"
+        """
+
+        def mainSourceFile = file("src/main/java/Main.java")
+        mainSourceFile.text = sourceFileWithGreeting("Hello World!")
+
+        when:
+        withRetentionRun "run"
+        then:
+        outputContains "Hello World!"
+        executedAndNotSkipped ":compileJava", ":classes", ":run"
+
+        when:
+        mainSourceFile.text = sourceFileWithGreeting("Hello VFS!")
+        withoutRetentionRun "run"
+        then:
+        outputContains "Hello VFS!"
+        executedAndNotSkipped ":compileJava", ":classes", ":run"
+    }
+
+    private void withRetentionRun(String... tasks) {
```

Comment:
How about you use a similar method as we use for enabling the build cache:
https://github.com/gradle/gradle/blob/b61074aeb3fb4918626fabcab3d89fba1c8eb43f/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy#L239-L242

So something like this would work:
```
withRentention().run "task"
```

WDYT?

Reply:
Good idea, fixed.